internal/cmd/agent/controller.go

```diff
@@ -2,109 +2,245 @@ package agent
 	"github.com/rancher/fleet/internal/cmd/agent/register"
+	"github.com/rancher/fleet/internal/cmd/agent/trigger"
+	"github.com/rancher/fleet/internal/helmdeployer"
+	"github.com/rancher/fleet/internal/manifest"
+	"github.com/rancher/fleet/pkg/apis/fleet.cattle.io/v1alpha1"
 
-	"github.com/rancher/lasso/pkg/mapper"
-	"github.com/rancher/wrangler/v2/pkg/kubeconfig"
-	"github.com/rancher/wrangler/v2/pkg/ratelimit"
-	"github.com/rancher/wrangler/v2/pkg/ticker"
+	"helm.sh/helm/v3/pkg/cli"
 
-	"k8s.io/apimachinery/pkg/api/meta"
-	"k8s.io/client-go/discovery"
-	"k8s.io/client-go/discovery/cached/memory"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/client-go/rest"
-	"k8s.io/client-go/restmapper"
-	"k8s.io/client-go/tools/clientcmd"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/healthz"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	metricsserver "sigs.k8s.io/controller-runtime/pkg/metrics/server"
+)
+
+var (
+	scheme      = runtime.NewScheme()
+	localScheme = runtime.NewScheme()
 )
 
 // defaultNamespace is the namespace to use for resources that don't specify a namespace, e.g. "default"
 const defaultNamespace = "default"
 
+func init() {
+	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
+	utilruntime.Must(v1alpha1.AddToScheme(scheme))
+	//+kubebuilder:scaffold:scheme
+
+	utilruntime.Must(clientgoscheme.AddToScheme(localScheme))
+}
+
 // start the fleet agent
-func start(ctx context.Context, kubeConfig, namespace, agentScope string) error {
-	clientConfig := kubeconfig.GetNonInteractiveClientConfig(kubeConfig)
-	kc, err := clientConfig.ClientConfig()
+// systemNamespace is the namespace the agent is running in, e.g. cattle-fleet-system
+func start(ctx context.Context, localConfig *rest.Config, systemNamespace, agentScope string) error {
+	// Registration is done in an init container. If we are here, we are already registered.
+	// Retrieve the existing config from the registration.
+	// Cannot start without kubeconfig for upstream cluster:
+	upstreamConfig, fleetNamespace, clusterName, err := loadRegistration(ctx, systemNamespace, localConfig)
 	if err != nil {
+		setupLog.Error(err, "unable to load registration and start manager")
 		return err
 	}
 
-	agentInfo, err := register.Get(ctx, namespace, kc)
+	// Start manager for upstream cluster, we do not use leader election
+	setupLog.Info("listening for changes on upstream cluster", "cluster", clusterName, "namespace", fleetNamespace)
+
+	metricsAddr := ":8080"
+	probeAddr := ":8081"
+
+	mgr, err := ctrl.NewManager(upstreamConfig, ctrl.Options{
+		Scheme:                 scheme,
+		Metrics:                metricsserver.Options{BindAddress: metricsAddr},
+		HealthProbeBindAddress: probeAddr,
+		LeaderElection:         false,
```

Comment:
Are we sure we want to remove leader election? If that is the case, do we want to do it in gitjob and fleet controller as well?

We would be losing the ability to be highly available, and we would see some incosistent behaviour if someone manually scale the statefulset

Reply:
We already removed leader election in  https://github.com/rancher/fleet/pull/1905/files
We can add a follow up item to bring it back. We'll need to discuss how that works together with the upcoming [sharding story](https://github.com/rancher/fleet/issues/1740).